README.md
# pcsensor-temper
Simple C-program to query temperature from TEMPer-USB-devices. There's a
Munin plugin included too. This repository is prepared for Debian
package builds.

## Supported and tested devices
* `ID 0c45:7401 Microdia` ([M-ware ID7747](http://amzn.to/2fUFUXM))

## Build from source
```bash
sudo apt install git build-essential libusb-dev
git clone "https://github.com/froonix/pcsensor-temper.git"
cd "pcsensor-temper" && make all

# copy udev rule (udev directory)
# copy program (pcsensor file)
# ...
```
## Binary packages (experimental)
Please take a look at the [wiki](https://github.com/froonix/pcsensor-temper/wiki/Binary-Debian-Packages).

## See also
* [Source of this fork](https://github.com/padelt/pcsensor-temper)
* [Original v1.0.0](http://www.isp-sl.com/pcsensor-1.0.0.tgz)


debian/changelog
pcsensor-temper (1.1.1-1) stable; urgency=medium

  * New release.

 -- Christian Schrötter <cs@fnx.li>  Tue, 18 Apr 2017 20:51:12 +0200

pcsensor-temper (1.1.0-3) stable; urgency=medium

  * First build for Debian Jessie.

 -- Christian Schrötter <cs@fnx.li>  Thu, 24 Nov 2016 18:16:40 +0100


twitter/ttytter.sh
#!/bin/bash

# Legacy script to tweet temp values.
# Only a simple example from my sys.

LC_NUMERIC=C

typeset -i tse
tse=$(date +%s)
typeset -i tss
tss=$tse-3600

cd /var/lib/munin/Misc || exit 1

max=$(rrdtool graph /dev/null -s "$tss" -e "$tse" \
	DEF:v1=room1-temper_0-temp-g.rrd:42:MAX \
	DEF:v2=room2-temper_0-temp-g.rrd:42:MAX \
	PRINT:v1:MAX:%lf \
	PRINT:v2:MAX:%lf \
	| tail -n +2)
max=$(sort -n -r <<< "$max" | head -n 1)
max=$(printf "%.2f" "$max" | sed 's/\./,/')

min=$(rrdtool graph /dev/null -s "$tss" -e "$tse" \
	DEF:v1=room1-temper_0-temp-g.rrd:42:MIN \
	DEF:v2=room2-temper_0-temp-g.rrd:42:MIN \
	PRINT:v1:MIN:%lf \
	PRINT:v2:MIN:%lf \
	| tail -n +2)
min=$(sort -n <<< "$min" | head -n 1)
min=$(printf "%.2f" "$min" | sed 's/\./,/')

avg=$(rrdtool graph /dev/null -s "$tss" -e "$tse" \
	DEF:v1=room1-temper_0-temp-g.rrd:42:AVERAGE \
	DEF:v2=room2-temper_0-temp-g.rrd:42:AVERAGE \
	PRINT:v1:AVERAGE:%lf \
	PRINT:v2:AVERAGE:%lf \
	| tail -n +2 \
	| awk '{ total += $1; count++ } END { print total/count }')
avg=$(printf "%.2f" "$avg" | sed 's/\./,/')

if [[ "$max" != "" && "$min" != "" && "$avg" != "" ]]
then
	ttytter -ssl -keyf=-example -status="Temperaturwerte der letzten Stunde: $min / $avg / $max °C (min/avg/max)" &> /dev/null && exit 0
else
	echo "Data source ERROR!"
	exit 2
fi

echo "Twitter update FAILED!"
exit 1


tmp/pcsensor-plain.c
/*
 * pcsensor-plain.c by Christian Schrötter (c) 2012 (c@schroetter.co.at)
 * based in pcsensor.c by Philipp Adelt (c) 2012 (info@philipp.adelt.net)
 * based on Juan Carlos Perez (c) 2011 (cray@isp-sl.com)
 * based on Temper.c by Robert Kavaler (c) 2009 (relavak.com)
 * All rights reserved.
 *
 * Temper driver for linux. This program can be compiled either as a library
 * or as a standalone program (-DUNIT_TEST). The driver will work with some
 * TEMPer usb devices from RDing (www.PCsensor.com).
 *
 * This driver works with USB devices presenting ID 0c45:7401.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 * 
 * THIS SOFTWARE IS PROVIDED BY Philipp Adelt (and other contributors) ''AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL Philipp Adelt (or other contributors) BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 */



#include <usb.h>
#include <stdio.h>
#include <time.h>

#include <string.h>
#include <errno.h>
#include <signal.h> 
 
 
#define VERSION "1.0.0~cs1"
 
#define VENDOR_ID  0x0c45
#define PRODUCT_ID 0x7401
 
#define INTERFACE1 0x00
#define INTERFACE2 0x01
 
const static int reqIntLen=8;
const static int reqBulkLen=8;
const static int endpoint_Int_in=0x82; /* endpoint 0x81 address for IN */
const static int endpoint_Int_out=0x00; /* endpoint 1 address for OUT */
const static int endpoint_Bulk_in=0x82; /* endpoint 0x81 address for IN */
const static int endpoint_Bulk_out=0x00; /* endpoint 1 address for OUT */
const static int timeout=5000; /* timeout in ms */
 
const static char uTemperatura[] = { 0x01, 0x80, 0x33, 0x01, 0x00, 0x00, 0x00, 0x00 };
const static char uIni1[] = { 0x01, 0x82, 0x77, 0x01, 0x00, 0x00, 0x00, 0x00 };
const static char uIni2[] = { 0x01, 0x86, 0xff, 0x01, 0x00, 0x00, 0x00, 0x00 };

static int bsalir=1;
static int seconds=5;
static int formato=0;
static int calibration=0;


void bad(const char *why) {
        fprintf(stderr,"Fatal error> %s\n",why);
        exit(17);
}
 
 
usb_dev_handle *find_lvr_winusb();
 
void usb_detach(usb_dev_handle *lvr_winusb, int iInterface) {
        int ret;
 
	ret = usb_detach_kernel_driver_np(lvr_winusb, iInterface);
} 

usb_dev_handle* setup_libusb_access(int devicenum) {
     usb_dev_handle *lvr_winusb;
    usb_set_debug(0);
     usb_init();
     usb_find_busses();
     usb_find_devices();
             
 
     if(!(lvr_winusb = find_lvr_winusb(devicenum))) {
                printf("Couldn't find the USB device, Exiting\n");
                return NULL;
        }
        
        
        usb_detach(lvr_winusb, INTERFACE1);
        

        usb_detach(lvr_winusb, INTERFACE2);
        
 
        if (usb_set_configuration(lvr_winusb, 0x01) < 0) {
                printf("Could not set configuration 1\n");
                return NULL;
        }
 

        // Microdia tiene 2 interfaces
        if (usb_claim_interface(lvr_winusb, INTERFACE1) < 0) {
                printf("Could not claim interface\n");
                return NULL;
        }
 
        if (usb_claim_interface(lvr_winusb, INTERFACE2) < 0) {
                printf("Could not claim interface\n");
                return NULL;
        }
 
        return lvr_winusb;
}
 
 
 
usb_dev_handle *find_lvr_winusb(int devicenum) {
        // iterates to the devicenum'th device for installations with multiple sensors
        struct usb_bus *bus;
        struct usb_device *dev;
 
        for (bus = usb_busses; bus; bus = bus->next) {
        for (dev = bus->devices; dev; dev = dev->next) {
                        if (dev->descriptor.idVendor == VENDOR_ID && 
                                dev->descriptor.idProduct == PRODUCT_ID ) {
                                if (devicenum>0) {
                                  devicenum--;
                                  continue;
                                }
                                usb_dev_handle *handle;
 
                                if (!(handle = usb_open(dev))) {
                                        printf("Could not open USB device\n");
                                        return NULL;
                                }
                                return handle;
                        }
                }
        }
        return NULL;
}
 
 
void ini_control_transfer(usb_dev_handle *dev) {
    int r,i;

    char question[] = { 0x01,0x01 };

    r = usb_control_msg(dev, 0x21, 0x09, 0x0201, 0x00, (char *) question, 2, timeout);
    if( r < 0 )
    {
          perror("USB control write"); bad("USB write failed"); 
    }


}
 
void control_transfer(usb_dev_handle *dev, const char *pquestion) {
    int r,i;

    char question[reqIntLen];
    
    memcpy(question, pquestion, sizeof question);

    r = usb_control_msg(dev, 0x21, 0x09, 0x0200, 0x01, (char *) question, reqIntLen, timeout);
    if( r < 0 )
    {
          perror("USB control write"); bad("USB write failed"); 
    }

}

void interrupt_transfer(usb_dev_handle *dev) {
 
    int r,i;
    char answer[reqIntLen];
    char question[reqIntLen];
    for (i=0;i<reqIntLen; i++) question[i]=i;
    r = usb_interrupt_write(dev, endpoint_Int_out, question, reqIntLen, timeout);
    if( r < 0 )
    {
          perror("USB interrupt write"); bad("USB write failed"); 
    }
    r = usb_interrupt_read(dev, endpoint_Int_in, answer, reqIntLen, timeout);
    if( r != reqIntLen )
    {
          perror("USB interrupt read"); bad("USB read failed"); 
    }
 
    usb_release_interface(dev, 0);
}

void interrupt_read(usb_dev_handle *dev) {
 
    int r,i;
    unsigned char answer[reqIntLen];
    bzero(answer, reqIntLen);
    
    r = usb_interrupt_read(dev, 0x82, answer, reqIntLen, timeout);
    if( r != reqIntLen )
    {
          perror("USB interrupt read"); bad("USB read failed"); 
    }

}

void interrupt_read_temperatura(usb_dev_handle *dev, float *tempC) {
 
    int r,i, temperature;
    unsigned char answer[reqIntLen];
    bzero(answer, reqIntLen);
    
    r = usb_interrupt_read(dev, 0x82, answer, reqIntLen, timeout);
    if( r != reqIntLen )
    {
          perror("USB interrupt read"); bad("USB read failed"); 
    }


    
    temperature = (answer[3] & 0xFF) + (answer[2] << 8);
    temperature += calibration;
    *tempC = temperature * (125.0 / 32000.0);

}

void bulk_transfer(usb_dev_handle *dev) {
 
    int r,i;
    char answer[reqBulkLen];

    r = usb_bulk_write(dev, endpoint_Bulk_out, NULL, 0, timeout);
    if( r < 0 )
    {
          perror("USB bulk write"); bad("USB write failed"); 
    }
    r = usb_bulk_read(dev, endpoint_Bulk_in, answer, reqBulkLen, timeout);
    if( r != reqBulkLen )
    {
          perror("USB bulk read"); bad("USB read failed"); 
    }


 
    usb_release_interface(dev, 0);
}
 

void ex_program(int sig) {
      bsalir=1;
 
      (void) signal(SIGINT, SIG_DFL);
}
 
int main( int argc, char **argv) {
 
     usb_dev_handle *lvr_winusb = NULL;
     float tempc;
     int c;
     struct tm *local;
     time_t t;
     int devicenum = 0;

     while ((c = getopt (argc, argv, "mfcvhn:l::a:")) != -1)
     switch (c)
       {
       case 'n':
         if (optarg != NULL) {
           if (!sscanf(optarg,"%i",&devicenum)==1) {
             fprintf (stderr, "Error: '%s' is not numeric.\n", optarg);
             exit(EXIT_FAILURE);
           }
         }
         break;
       case 'c':
         formato=1; //Celsius
         break;
       case 'f':
         formato=2; //Fahrenheit
         break;
       case 'l':
         if (optarg!=NULL){
           if (!sscanf(optarg,"%i",&seconds)==1) {
             fprintf (stderr, "Error: '%s' is not numeric.\n", optarg);
             exit(EXIT_FAILURE);
           } else {           
              bsalir = 0;
              break;
           }
         } else {
           bsalir = 0;
           seconds = 5;
           break;
         }
       case 'a':
         if (!sscanf(optarg,"%i",&calibration)==1) {
             fprintf (stderr, "Error: '%s' is not numeric.\n", optarg);
             exit(EXIT_FAILURE);
         } else {           
              break;
         }
       case '?':
       case 'h':
         printf("pcsensor-plain version %s\n",VERSION);
	 printf("      Aviable options:\n");
	 printf("          -h help\n");
	 printf("          -n[i] use device number i (0 is the first one found on the bus)\n");
	 printf("          -l[n] loop every 'n' seconds, default value is 5s\n");
	 printf("          -c output only in Celsius\n");
	 printf("          -f output only in Fahrenheit\n");
	 printf("          -a[n] increase or decrease temperature in 'n' degrees for device calibration\n");
  
	 exit(EXIT_FAILURE);
       default:
         if (isprint (optopt))
           fprintf (stderr, "Unknown option `-%c'.\n", optopt);
         else
           fprintf (stderr,
                    "Unknown option character `\\x%x'.\n",
                    optopt);
         exit(EXIT_FAILURE);
       }

     if (optind < argc) {
        fprintf(stderr, "Non-option ARGV-elements, try -h for help.\n");
        exit(EXIT_FAILURE);
     }
 
     if ((lvr_winusb = setup_libusb_access(devicenum)) == NULL) {
         exit(EXIT_FAILURE);
     } 

     (void) signal(SIGINT, ex_program);

     ini_control_transfer(lvr_winusb);
      
     control_transfer(lvr_winusb, uTemperatura );
     interrupt_read(lvr_winusb);
 
     control_transfer(lvr_winusb, uIni1 );
     interrupt_read(lvr_winusb);
 
     control_transfer(lvr_winusb, uIni2 );
     interrupt_read(lvr_winusb);
     interrupt_read(lvr_winusb);


 
     do {
           control_transfer(lvr_winusb, uTemperatura );
           interrupt_read_temperatura(lvr_winusb, &tempc);

              if (formato==2) {
                  printf("%.2f\n", (9.0 / 5.0 * tempc + 32.0));
              } else if (formato==1) {
                  printf("%.2f\n", tempc);
              } else {
                  fprintf (stderr, "Please use -c/-f.\n", optopt);
                  exit(EXIT_FAILURE);
              }
           
           if (!bsalir)
              sleep(seconds);
     } while (!bsalir);
                                       
     usb_release_interface(lvr_winusb, INTERFACE1);
     usb_release_interface(lvr_winusb, INTERFACE2);
     
     usb_close(lvr_winusb); 
      
     return 0; 
}


pcsensor.c
/*
 * PCSENSOR-TEMPER
 *
 * Temper driver for linux. This program can be compiled either as a
 * library or as a standalone program (-DUNIT_TEST). The driver will
 * work with some TEMPer usb devices from RDing (www.PCsensor.com).
 *
 * This driver works with USB devices presenting ID 0c45:7401.
 *
 * ---------------------------------------------------------------------
 *
 * Copyright (c) 2016, Christian Schrötter <cs@fnx.li>
 * Copyright (c) 2012, Philipp Adelt <info@philipp.adelt.net>
 * Copyright (c) 2011, Juan Carlos Perez <cray@isp-sl.com>
 * Copyright (c) 2009, Robert Kavaler <relavak.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <usb.h>
#include <stdio.h>
#include <time.h>
#include <string.h>
#include <errno.h>
#include <signal.h>
#include <ctype.h>

#define VERSION "1.1.1"

#define VENDOR_ID  0x0c45
#define PRODUCT_ID 0x7401

#define INTERFACE1 0x00
#define INTERFACE2 0x01

const static int reqIntLen  = 8;
const static int reqBulkLen = 8;

const static int timeout           = 5000; /* timeout in ms                */
const static int endpoint_Int_in   = 0x82; /* endpoint 0x81 address for IN */
const static int endpoint_Int_out  = 0x00; /* endpoint 1 address for OUT   */
const static int endpoint_Bulk_in  = 0x82; /* endpoint 0x81 address for IN */
const static int endpoint_Bulk_out = 0x00; /* endpoint 1 address for OUT   */

const static char uTemp[] = { 0x01, 0x80, 0x33, 0x01, 0x00, 0x00, 0x00, 0x00 };
const static char uIni1[] = { 0x01, 0x82, 0x77, 0x01, 0x00, 0x00, 0x00, 0x00 };
const static char uIni2[] = { 0x01, 0x86, 0xff, 0x01, 0x00, 0x00, 0x00, 0x00 };

static int bsalir      = 1;
static int debug       = 0;
static int seconds     = 5;
static int format      = 0;
static int mrtg        = 0;
static int calibration = 0;

void bad(const char *why)
{
	fprintf(stderr, "Fatal error> %s\n", why);
	exit(17);
}

usb_dev_handle *find_lvr_winusb();

void usb_detach(usb_dev_handle *lvr_winusb, int iInterface)
{
	int ret;

	ret = usb_detach_kernel_driver_np(lvr_winusb, iInterface);

	if(debug)
	{
		if(ret)
		{
			if(errno == ENODATA)
			{
				printf("Device already detached\n");
			}
			else
			{
				printf("Detach failed: %s[%d]\n", strerror(errno), errno);
				printf("Continuing anyway\n");
			}
		}
		else
		{
			printf("detach successful\n");
		}
	}
}

usb_dev_handle* setup_libusb_access(int devicenum)
{
	usb_dev_handle *lvr_winusb;

	if(debug)
	{
		usb_set_debug(255);
	}
	else
	{
		usb_set_debug(0);
	}
	usb_init();
	usb_find_busses();
	usb_find_devices();

	if(!(lvr_winusb = find_lvr_winusb(devicenum)))
	{
		printf("Couldn't find the USB device, Exiting\n");
		return NULL;
	}

	usb_detach(lvr_winusb, INTERFACE1);
	usb_detach(lvr_winusb, INTERFACE2);

	if(usb_reset(lvr_winusb) != 0 /*&& debug*/)
	{
		printf("Could not reset device.\n");
		return NULL;
	}

	if(usb_set_configuration(lvr_winusb, 0x01) < 0)
	{
		printf("Could not set configuration 1\n");
		return NULL;
	}

	if(usb_claim_interface(lvr_winusb, INTERFACE1) < 0 || usb_claim_interface(lvr_winusb, INTERFACE2) < 0)
	{
		printf("Could not claim interface\n");
		return NULL;
	}

	return lvr_winusb;
}

usb_dev_handle *find_lvr_winusb(int devicenum)
{
	struct usb_bus *bus;
	struct usb_device *dev;

	for(bus = usb_busses; bus; bus = bus->next)
	{
		for(dev = bus->devices; dev; dev = dev->next)
		{
			if(dev->descriptor.idVendor == VENDOR_ID && dev->descriptor.idProduct == PRODUCT_ID )
			{
				if (devicenum>0)
				{
					devicenum--;
					continue;
				}

				usb_dev_handle *handle;
				if(debug)
				{
					printf("lvr_winusb with Vendor Id: %x and Product Id: %x found.\n", VENDOR_ID, PRODUCT_ID);
				}

				if(!(handle = usb_open(dev)))
				{
					printf("Could not open USB device\n");
					return NULL;
				}

				return handle;
			}
		}
	}

	return NULL;
}

void ini_control_transfer(usb_dev_handle *dev)
{
	int r, i;
	char question[] = { 0x01, 0x01 };

	r = usb_control_msg(dev, 0x21, 0x09, 0x0201, 0x00, (char *) question, 2, timeout);
	if( r < 0 )
	{
		perror("USB control write");
		bad("USB write failed");
	}

	if(debug)
	{
		for(i = 0 ; i < reqIntLen; i++)
		{
			printf("%02x ", question[i] & 0xFF);
		}
		printf("\n");
	}
}

void control_transfer(usb_dev_handle *dev, const char *pquestion)
{
	int r, i;
	char question[reqIntLen];

	memcpy(question, pquestion, sizeof question);

	r = usb_control_msg(dev, 0x21, 0x09, 0x0200, 0x01, (char *) question, reqIntLen, timeout);
	if(r < 0)
	{
		perror("USB control write");
		bad("USB write failed");
	}

	if(debug)
	{
		for(i = 0; i < reqIntLen; i++)
		{
			printf("%02x ", question[i] & 0xFF);
		}
		printf("\n");
	}
}

void interrupt_transfer(usb_dev_handle *dev)
{
	int r, i;
	char answer[reqIntLen];
	char question[reqIntLen];

	for(i = 0; i < reqIntLen; i++)
	{
		question[i] = i;
	}

	r = usb_interrupt_write(dev, endpoint_Int_out, question, reqIntLen, timeout);
	if(r < 0)
	{
		perror("USB interrupt write");
		bad("USB write failed");
	}

	r = usb_interrupt_read(dev, endpoint_Int_in, answer, reqIntLen, timeout);
	if(r != reqIntLen)
	{
		perror("USB interrupt read");
		bad("USB read failed");
	}

	if(debug)
	{
		for(i = 0; i < reqIntLen; i++)
		{
			printf("%i, %i, \n", question[i], answer[i]);
		}
	}

	usb_release_interface(dev, 0);
}

void interrupt_read(usb_dev_handle *dev)
{
	int r, i;
	unsigned char answer[reqIntLen];
	bzero(answer, reqIntLen);

	r = usb_interrupt_read(dev, 0x82, answer, reqIntLen, timeout);
	if(r != reqIntLen)
	{
		perror("USB interrupt read");
		bad("USB read failed");
	}

	if(debug)
	{
		for(i = 0; i < reqIntLen; i++)
		{
			printf("%02x ", answer[i] & 0xFF);
		}
		printf("\n");
	}
}

void interrupt_read_temperature(usb_dev_handle *dev, float *tempC)
{
	int r, i, temperature;
	unsigned char answer[reqIntLen];
	bzero(answer, reqIntLen);

	r = usb_interrupt_read(dev, 0x82, answer, reqIntLen, timeout);
	if(r != reqIntLen)
	{
		perror("USB interrupt read");
		bad("USB read failed");
	}

	if(debug)
	{
		for (i = 0; i < reqIntLen; i++)
		{
			printf("%02x ", answer[i] & 0xFF);
		}
		printf("\n");
	}

	// temperature = (answer[3] & 0xFF) + (answer[2] << 8);
	temperature = (answer[3] & 0xFF) + ((signed char) answer[2] << 8);
	temperature += calibration;
	*tempC = temperature * (125.0 / 32000.0);
}

void bulk_transfer(usb_dev_handle *dev)
{
	int r, i;
	char answer[reqBulkLen];

	r = usb_bulk_write(dev, endpoint_Bulk_out, NULL, 0, timeout);
	if(r < 0)
	{
		perror("USB bulk write");
		bad("USB write failed");
	}

	r = usb_bulk_read(dev, endpoint_Bulk_in, answer, reqBulkLen, timeout);
	if(r != reqBulkLen)
	{
		perror("USB bulk read");
		bad("USB read failed");
	}

	if(debug)
	{
		for(i = 0; i < reqBulkLen; i++)
		{
			printf("%02x ", answer[i] & 0xFF);
		}
	}

	usb_release_interface(dev, 0);
}

void ex_program(int sig)
{
	bsalir=1;
	(void) signal(SIGINT, SIG_DFL);
}

void usage(const char* progname)
{
	printf(
		"pcsensor version %s\n"
		"USAGE: %s [options] <arguments>\n"
		"  ARGUMENTS:\n"
		"  OPTIONS:\n"
		"    -h help\n"
		"    -v verbose\n"
		"    -n[i] use device number i (0 is the first one found on the bus)\n"
		"    -l[n] loop every 'n' seconds, default value is 5\n"
		"    -c output only in Celsius\n"
		"    -f output only in Fahrenheit\n"
		"    -a[n] increase or decrease temperature in 'n' degrees for device calibration\n"
		"    -m output for mrtg integration\n"
		"    -p output only the temperature value\n"
			, VERSION
			, progname
	);
}

int main( int argc, char **argv)
{
	usb_dev_handle *lvr_winusb = NULL;
	float tempc;
	float tempf;
	int c;
	struct tm *local;
	time_t t;
	int devicenum = 0;
	int plain = 0;

	while((c = getopt(argc, argv, "mfcvhpn:l::a:")) != -1)
	{
		switch (c)
		{
			case 'v':
				debug = 1;
				break;

			case 'n':
				if(optarg != NULL)
				{
					if(!sscanf(optarg, "%i", &devicenum) == 1)
					{
						fprintf (stderr, "Error: '%s' is not numeric.\n", optarg);
						exit(EXIT_FAILURE);
					}
				}
				break;

			case 'c':
				//Celsius
				format = 1;
				break;

			case 'f':
				//Fahrenheit
				format = 2;
				break;

			case 'm':
				mrtg = 1;
				break;

			case 'p':
				// raw output mode
				plain = 1;
				break;

			case 'l':
				if(optarg!=NULL)
				{
					if(!sscanf(optarg, "%i", &seconds) == 1)
					{
						fprintf (stderr, "Error: '%s' is not numeric.\n", optarg);
						exit(EXIT_FAILURE);
					}
					else
					{
						bsalir = 0;
					}
				}
				else
				{
					bsalir = 0;
					seconds = 5;
				}
				break;

			case 'a':
				if(!sscanf(optarg, "%i", &calibration) == 1)
				{
					fprintf (stderr, "Error: '%s' is not numeric.\n", optarg);
					exit(EXIT_FAILURE);
				}
				break;

			case '?':
			case 'h':
				usage(argv[0]);
				exit(EXIT_FAILURE);

			default:
				if(isprint(optopt))
				{
					fprintf (stderr, "Unknown option `-%c'.\n", optopt);
				}
				else
				{
					fprintf (stderr, "Unknown option character `\\x%x'.\n", optopt);
				}
				exit(EXIT_FAILURE);
				break;
		}
	}

	if(optind < argc)
	{
		fprintf(stderr, "Non-option ARGV-elements, try -h for help.\n");
		exit(EXIT_FAILURE);
	}

	if((lvr_winusb = setup_libusb_access(devicenum)) == NULL)
	{
		exit(EXIT_FAILURE);
	}

	(void) signal(SIGINT, ex_program);

	ini_control_transfer(lvr_winusb);

	control_transfer(lvr_winusb, uTemp );
	interrupt_read(lvr_winusb);

	control_transfer(lvr_winusb, uIni1 );
	interrupt_read(lvr_winusb);

	control_transfer(lvr_winusb, uIni2 );
	interrupt_read(lvr_winusb);
	interrupt_read(lvr_winusb);

	do
	{
		control_transfer(lvr_winusb, uTemp);
		interrupt_read_temperature(lvr_winusb, &tempc);
		tempf = 9.0 / 5.0 * tempc + 32.0;

		t = time(NULL);
		local = localtime(&t);

		if(mrtg)
		{
			if(format == 2)
			{
				printf("%.2f\n", tempf);
				printf("%.2f\n", tempf);
			}
			else
			{
				printf("%.2f\n", tempc);
				printf("%.2f\n", tempc);
			}

			printf("%02d:%02d\n", local->tm_hour, local->tm_min);
			printf("pcsensor\n");
		}
		else if(plain)
		{
			if(format == 2)
			{
				printf("%.2f\n", tempf);
			}
			else if(format == 1)
			{
				printf("%.2f\n", tempc);
			}
			else
			{
				fprintf(stderr, "Please use -c/-f.\n");
				exit(EXIT_FAILURE);
			}
		}
		else
		{
			printf("%04d/%02d/%02d %02d:%02d:%02d ", local->tm_year +1900, local->tm_mon + 1, local->tm_mday, local->tm_hour, local->tm_min, local->tm_sec);

			if(format == 2)
			{
				printf("Temperature %.2fF\n", tempf);
			}
			else if(format == 1)
			{
				printf("Temperature %.2fC\n", tempc);
			}
			else
			{
				printf("Temperature %.2fF %.2fC\n", tempf, tempc);
			}
		}

		if(!bsalir)
		{
			sleep(seconds);
		}
	}
	while(!bsalir);

	usb_release_interface(lvr_winusb, INTERFACE1);
	usb_release_interface(lvr_winusb, INTERFACE2);

	usb_close(lvr_winusb);

	return 0;
}
